implement get process document validator

LBHTKarki opened the pull request:
Implement validator for get process data

Duslerke committed:
Modified the GetProcessImageRequest Validator.

Added 2 extra tests for the 2 new fields of GetProcessImageRequest object ("processType, fileExtension). These tests are about throwing an error when a blank parameter is supplied. Added implementaion to validator based of these tests.

Also added back the removed tests and validation for the "processRef" and "imageId" fields. The ones that did the not null, not empty checks. Reason for it is that even though request with such null values wouldn't pass the validation, it wouldn't give the error message specific enough to what the issue is. I feel that providing 'null' argument versus an invalid one should return different error messages.

Now, there was an argument raised on the PR against these checks saying that: since now the parameters are being passed in through url path, when you don't provide them, it will simply be interpreted as a different path - hence 404 path not found.
However, in response to that, I'd say: if you had provided a url encoded whitespace (which sometimes gets encoded automatically) as an argument, it would count as a valid path, and the  .NET would interpret the whitespace as 'null'. I checked it through the browser and debugger - it confirms it.

Example: "{insert your host}/api/v1/processImageData/%20/a68e234a-bd08-3542-4b89-42572b6adc65/e68e234a-bd08-3542-4b89-42572b6adc65/%20"

Duslerke committed:
Added regions to MatProcessDataHelper for easier navigation.

Also changed class to static, since it contained nothing but static methods anyway. There's no need to be false hinting that you might need an instance of it.

Duslerke committed:
Renamed the GetProcessImage endpoint Controller tests to be more desciptive about what they actually test.

-------------- Request unimportant --------------

In a lot of cases where the test names starts with: "given_a_valid_request" or "given_an_invalid_request". The request object created or referenced in the name does not matter at all. Example:
Ln189 "given_a_valid_request_when_GetProcessImage_controller_method_is_called_then_it_returns_a_200_Ok_response"
this test name was misleading because due to this being a unit test - all the dependencies of class under test (in this case controller) are being mocked. The mock itself contains no logic apart from whatever you make it return when doing mock setup in the arrange step. So in this test's case it means that it does not matter what request I pass into it, it will always return whatever I specified that mock to return. So, if I'm testing "given_a_valid_request", what it actually means is that I'm testing whether whatever is specified happens based on request validation logic returning success. So the test is more like "under_the_validation_control_flow_of_success_X_happens".

To make this explicit, I renamed the test as well as moved empty object creation directly into argument place in the controller call (act phase). To signify that the request itself is not part of the test context, hence shouldn't be in the arrange step.

There are many more cases like this, so I won't cover them.

------------------Status Codes------------------

Ln263 status code wasn't even being used, because the test didn't even check for that - that was the job for the test before it. It's good to have these tests separated for easier and quicked troubleshooting.

Ln289 Since the test states that it's checking for 500 status code result in its name, this means that it should probably be put in the arrange step to make this more explicit.

--------------------"Correct Data"--------------------

I figured that it's a pretty bad description for whatever you're expecting as a result from the test. This is because this 'term' is too abstract and can mean too many things. So I changed this term into something more specific accross the Get endpoint tests in order to clarify what we're actually testing for.

----------------Modified few logger tests----------------

Tests in lines: "Ln423" -> "Ln411" and "Ln445" -> "Ln431" were indeed testing what they said, however they did not reflect the intention behind the tests.

The point of those tests was to show that logger is being called regardless of validation controll flow - as in it was called prior to the validation, however all they were testing was that the logger was being called at least once accross the entire validation controll flow.

So I changed these test to test exactly what they meant - that the request gets registered before it gets validated, meaning this registering action does not depend on the validation result.

To do this I had to ensure there's no Validator mock setup, which would influence the decision to call logger or not. This of course means a crash within the controller, since it's trying to make a decision based on null result now, but this crash can get mitigated by putting an empty try-catch block to ignore the error.

The whole idea is:
- If logger call depends on the validation result, due to a crash (and absense of control flow decision) in the controller, the call  to logger will never happen. And the test will fail due to there being 0 calls to logger in the assert step.
- However, if the logger call is independant of validation controll flow, then we should expect the call to happen prior to validation (prior to the error), and the test passes.

Now the counter argument could be: "what if the logger get called after all the validation, or in every branch of validation decision?". I'd say that even under these circumstances, you're still dependent on the validation - maybe not on the decision it makes, but on it not crashing, which is a bad way to ensure that every request coming into the controller gets registered.

Duslerke committed:
Modified GetProcessImage logger tests and logger.

Modified the tests to expect 4 properties logged from the GetProcessImageRequest object. (Previously it was only 2, only after that additional 2 were added.)

Added the implementation to make modified tests pass.

Modified couple names of the logger tests again to be more explicit that this logger call does not depend not only on the result of the validation, but the behavior (like crashing) as well.

Added additional test that ensures that the initial logger call will also have null values from the request replaced with "null" strings. (In order to have some consistency accross the calls; Also helps with troubleshooting )

LBHTKarki committed:
implement get process document validator

LBHTKarki committed:
implement get process document validator

MirelaGeorgiou commented on mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:
```diff
@@ -73,12 +69,114 @@ public void given_a_processRef_when_getprocessdata_method_is_called_it_then_call
+            Assert.NotNull(result);
+            Assert.NotNull(resultContent);
+            Assert.AreEqual(errorResponse, resultContent);
+            Assert.AreEqual(200, result.StatusCode);
```
We should change that to 404 at some point to keep it consistent with other endpoints

LBHTKarki replied:
Changed to 404

MirelaGeorgiou commented on mat-process-api.Tests/V1/Validators/GetProcessRequestValidatorTest.cs:
```diff
@@ -0,0 +1,72 @@
+
+        [Test]
+        public void given_whitespace_processRef_the_validator_should_return_an_error()
+        {
```
Those could also be made into a single test with three test cases

LBHTKarki replied:
Combined to single test

LBHTKarki replied:
New test coverage in latest commit to check for error messages as well

MirelaGeorgiou commented on mat-process-api.Tests/V1/Validators/GetProcessRequestValidatorTest.cs:
```diff
@@ -0,0 +1,72 @@
+        }
+
+        [Test]
+        public void given_valid_guid_string_the_validator_should_return_an_error()
```
Test name misleading - it states that the validator should return an error, when it shouldn't

LBHTKarki replied:
Renamed the test

MirelaGeorgiou commented on mat-process-api/V1/Controllers/ProcessDataController.cs:
```diff
@@ -45,12 +48,28 @@ public class ProcessDataController : Controller
+            {
+                try
+                {
+                    _logger.LogInformation($"Get ProcessData request for process ID {request.processRef}");
```
The logger should come before the validation call so we have a log that the request for process ref X has occured

LBHTKarki replied:
Moved logger to run first

MirelaGeorgiou commented on mat-process-api/mat-process-api.csproj:
```diff
@@ -12,6 +12,9 @@
       <DebugType>full</DebugType>
       <DebugSymbols>true</DebugSymbols>
     </PropertyGroup>
+    <ItemGroup>
```
I believe this should not be added

LBHTKarki replied:
Deleted unnecessary lines. Leftover from moving files earlier

MirelaGeorgiou reviewed (changes requested):
Please look at the comments. Also, the validator needs to be registered in the Startup class for DI.

LBHTKarki committed:
Resolve issues highlighted in PR27

LBHTKarki committed:
Resolve issues highlighted in PR27

LBHTKarki commented:
Resolved all issues, including the validator registration, with new commit

Duslerke committed:
Modified GetProcessRequestValidator, made it return different error messages.

The validator now returns different error messages based on whether the field was null or empty. Pretty handy for whoever is consuming or trying to connect to the API.

Duslerke committed:
Modified ErrorThrowerHelper to return Exceptions rahter than throw them.

Motivation behind it is easier to read tests, as you no longer have to do the .Returns(()=> ...) method to throw an exception. Also no longer need the pointless return within that same Return to keep compiler happy.

Another bonus is that tests setup is much simpler looking than before making test easier to read and understand.

The helper method also benefits from this by not needing parameters anymore. The sole reason it had parameters in before was so that the same error could be thrown twice. Now it's done, by returning random error, and seting up its throw.

Duslerke committed:
Improved ErrorThrowerHelper dual-layer base error messages.

So that they would be easier to distinguish if debugger use is needed.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:
```diff
@@ -73,12 +69,114 @@ public void given_a_processRef_when_getprocessdata_method_is_called_it_then_call
+        }
+
+        [Test]
+        public void test_that_controller_returns_correct_message_and_status_code_when_no_document_was_found()
```
More specific test name would be good. Like:
`given_DocumentNotFound_exception_is_thrown_when_GetProcessData_controller_method_returns_404_status_code_with_correct_error_message`

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:
```diff
@@ -73,12 +69,114 @@ public void given_a_processRef_when_getprocessdata_method_is_called_it_then_call
+                processRef = faker.Random.Guid().ToString()
+            };
+
+            _mockUsecase.Setup(x => x.ExecuteGet(request)).Throws<AggregateException>();
```
Triangulation improvement possible after the merge though...

LBHTKarki replied:
Updated to use ErrorThrowerHelper

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:
```diff
@@ -73,12 +69,114 @@ public void given_a_processRef_when_getprocessdata_method_is_called_it_then_call
+        }
+
+        [Test]
+        public void test_that_get_controller_returns_correct_500_status_code_when_an_error_has_occured()
```
Possible test name improvement:
`given_an_unexpected_exception_thrown_when_xxxx_controller_method_is_called_then_controller_returns_500_status_code_response`

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:
```diff
@@ -73,12 +69,114 @@ public void given_a_processRef_when_getprocessdata_method_is_called_it_then_call
+        }
+
+        [Test]
+        public void test_that_get_controller_returns_correct_500_status_code_when_an_error_has_occured()
```
I think it's better to split this test into 2:
1) checks the response type and code
2) checks the contents (message)

LBHTKarki replied:
Split this into two

Duslerke commented on mat-process-api.Tests/V1/Validators/GetProcessRequestValidatorTest.cs:
```diff
@@ -0,0 +1,53 @@
+            validator = new GetProcessDocumentRequestValidator();
+        }
+
+        [Test]
```
Don't need [test] when you have [TestCase]

Duslerke commented on mat-process-api.Tests/V1/Validators/GetProcessRequestValidatorTest.cs:
```diff
@@ -0,0 +1,53 @@
+        [TestCase(null)]
+        [TestCase("")]
+        [TestCase(" ")]
+        public void given_null_or_empty_or_whitespace_processRef_the_validator_should_return_an_error(string processRef)
```
If you will want to test error messages you'll have to split 'null' test into separate [Test].

Duslerke commented on mat-process-api.Tests/V1/Validators/GetProcessRequestValidatorTest.cs:
```diff
@@ -0,0 +1,53 @@
+        }
+
+        [Test]
+        public void given_invalid_guid_string_the_validator_should_return_an_error()
```
Need more test cases. Something like [that](https://github.com/LBHackney-IT/mat-process-api/blob/16a76fdc79d75180c9e51c419a93945fd607a0e9/mat-process-api.Tests/V1/Validators/PostProcessImageRequestValidatorTests.cs#L84).

Duslerke commented on mat-process-api.Tests/V1/Validators/GetProcessRequestValidatorTest.cs:
```diff
@@ -0,0 +1,53 @@
+        }
+
+        [Test]
+        public void given_valid_guid_string_the_validator_should_not_return_an_error()
```
Test name improvement - specify Validator name and property name.

Duslerke reviewed (approved):
Looks all good. Just bit tidying would be nice.

Duslerke committed:
Clarified the GetProcessImage endpoint initial logger call test.

Had to use summary not to get limited by trying to make single line comment at sane levels of length.

Had to clarify this bit better, due to previous explanation not being clear enough, and bit confusing in regards to "making validation crash" bit.

Duslerke committed:
Merge pull request #29 from LBHackney-IT/feature/Validation-and-minor-Improvements

Feature/validation and minor improvements

Merge of Few extra validation rules (due to request object change), 2 extra fields of the request becoming logged by logger (due to request object change)  + few minor improvements in testing (to make tests clearer on what they test, and make their setup easier to follow an read)

LBHTKarki committed:
Improvements based on PR comments

LBHTKarki committed:
Improvements based on PR comments

LBHTKarki committed:
fix test name

LBHTKarki committed:
fix test name

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:
```diff
@@ -128,7 +128,7 @@ public void given_GetProcessData_controller_method_call_when_request_validation_
 
         [Test]
-        public void test_that_controller_returns_correct_message_and_status_code_when_no_document_was_found()
+        public void given_DocumentNotFound_exception_is_thrown_when_GetProcessData_controller_method_returns_404_status_code_with_correct_error_message()
```
There's an accidentally deleted bit in the test name:
`when_GetProcessData_controller_[deleted bit]_method_returns`

Duslerke commented on mat-process-api.Tests/V1/Validators/GetProcessRequestValidatorTest.cs:
```diff
@@ -19,28 +19,45 @@ public void SetUp()
         [TestCase("")]
         [TestCase(" ")]
-        public void given_null_or_empty_or_whitespace_processRef_the_validator_should_return_an_error(string processRef)
+        public void given_null_or_empty_or_whitespace_processRef_the_GetProcessDocumentRequestValidator_should_return_an_error(string processRef)
```
You don't need this test anymore, because you have it full covered with 2 tests bellow:
`given_null_processRef_the_GetProcessDocumentRequestValidator_should_return_a_correct_error_message` and
`given_empty_or_whitespace_processRef_the_GetProcessDocumentRequestValidator_should_return_a_correct_error_message`

LBHTKarki replied:
I've removed it as redundant

Duslerke commented on mat-process-api.Tests/V1/Validators/GetProcessRequestValidatorTest.cs:
```diff
@@ -19,28 +19,45 @@ public void SetUp()
+        [TestCase("6418aa75-f5ee6975-5279d7bcfee5f948")] //missing dashes
+        [TestCase("a98e73-4abd0-83542-4b73225-72b6adc65")] //dashes in the wrong places
+        [TestCase("ju5t70m3-w0rd-th4t-w1ll-m34nn0th1ng0")] //uses characters that are not among hexadecimal numerals
+        [TestCase("randomstring")]
```
Sorry, given you bad advice, now that I think a bit more, this 5th test case (`[TestCase("randomstring")]`) is not necessary because it falls under 2 categories tested above anyway. The categories are:
1) Less characters than it should contain
2) Uses characters not among Hex numerals
So it ends up testing things that were tested anyway. Better to take it out.

LBHTKarki replied:
I've removed that test case

Duslerke reviewed (approved):
Few more changes would be nice for the tests.

LBHTKarki committed:
remove reduntant  tests

LBHTKarki committed:
remove reduntant  tests

LBHTKarki committed:
remove unnecessary test case

LBHTKarki committed:
remove unnecessary test case

LBHTKarki committed:
Merge branch 'feature/implement-get-process-request-validator' of https://github.com/LBHackney-IT/mat-process-api into feature/implement-get-process-request-validator

LBHTKarki committed:
add test for error messages
